### ISSUE TITLE
Testing Errors - Failed to match 1 mock for this query. 

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-admin#555]
**Describe the bug**

1. We test all our code with each pull request. 
1. The tests pass as expected however we have noticed that they are giving excessive amounts of warnings that make it hard to identify when other tests fail.
1. You can see this problem with this log of the GitHub test run https://github.com/PalisadoesFoundation/talawa-admin/actions/runs/4359429650/jobs/7621306013
1.  Almost all of the give a warning with this string:

         This typically indicates a configuration error in your mocks setup, usually due to a typo or mismatched variable.

**To Reproduce**

1. Run tests
1. Many warnings are created

**Expected behavior**

1. No errors

**Actual behavior**
Repeated messages like this in the test logs:

```      
      This typically indicates a configuration error in your mocks setup, usually due to a typo or mismatched variable.
      Failed to match 1 mock for this query. The mocked response had the following variables:
        {}
      
      This typically indicates a configuration error in your mocks setup, usually due to a typo or mismatched variable.

      72 |     setSearchByName(value);
      73 |
    > 74 |     refetch({
         |     ^
      75 |       filter: searchByName,
      76 |     });
      77 |   };
```

**Screenshots**

- N/A

**Additional details**

- N/A

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by aviraldevv** [comment]
@palisadoes There are a lot of warnings generated which must have piled over a long time, we can create several child issues to fix them up. It will greatly improve the functionality of talawa-admin.

**Comment by palisadoes** [comment]
Thanks,

1. How should we proceed?
1. How many issues do you think it would take and what general categories? 

This looks like a very brittle situation and we may be fooling ourselves about the true stability of the app.

**Comment by aviraldevv** [comment]
@palisadoes give me some time I need to take a look at them.

**Comment by anshgoyalevil** [comment]
@palisadoes 

Providing some possible fixes for this issues. The following points need to be taken in consideration by the test writers:

 - You’re using the wrong query.
 - You don’t have enough mocked responses — you need to mock each query used in that component
- If you have variables, they are not in the format that is expected. Double-check how you’re passing variables.
- Your response is not in the correct format. This could be the type name, the response fields, etc. I suggest looking at the network call and copying the response.
- Data is not matching — make sure your mocked data is consistent between the query, mutation, or response
- *[Most Important]* The number of mocked responses needs to match the number of times your query is called — check on queries that refetch or fetch more.

Moreover, the test suite of Talawa Admin is incomplete. We should integrate some E2E tests into Talawa Admin Project.

E2E tests check the complete application stack, like a real user does. It uses Jest and Puppeteer libraries to create a headless browser testing state. We can test the complete stack (Talawa API, Admin) with E2E tests.

This should be included as a separate idea for upcoming GSoC, or should be merged into other ideas.

**Comment by aviraldevv** [comment]
@palisadoes I agree with @anshgoyalevil this issue seems strenous to tackle it will take time. I am still working on crunching the data and possible fixes.

**Comment by palisadoes** [comment]
At a minimum we should determine a way to make the tests fail when these types of errors occur. The testing regime is therefore flawed.

**Comment by anshgoyalevil** [comment]
> At a minimum we should determine a way to make the tests fail when these types of errors occur. The testing regime is therefore flawed.

Exactly, our test setup is allowing those warning. We should fail the test if those warning appear.

**Comment by anshgoyalevil** [comment]
> > @palisadoes I agree with @anshgoyalevil this issue seems strenous to tackle it will take time. I am still working on crunching the data and possible fixes.
> 
> Yes, but we also now have many Outreachy participants who would be eager to rectify this. They would be good first issues . Rectification could be quicker than you think.

Yes, the issue is simple. The test writer should rectify it during writing the tests.

For now, we can open several issues and allow outreachy applicants to rectify them. It's just due to improper mocks.

**Comment by aviraldevv** [comment]
> > @palisadoes I agree with @anshgoyalevil this issue seems strenous to tackle it will take time. I am still working on crunching the data and possible fixes.
> 
> Yes, but we also now have many Outreachy participants who would be eager to rectify this. They would be good first issues . Rectification could be quicker than you th

> > At a minimum we should determine a way to make the tests fail when these types of errors occur. The testing regime is therefore flawed.
> 
> Exactly, our test setup is allowing those warning. We should fail the test if those warning appear.

Actually we need to understand how react is build @anshgoyalevil . Mismatching of variable for same purpose is what it not permits but just throws a warning.

**Comment by aviraldevv** [comment]
> > > @palisadoes I agree with @anshgoyalevil this issue seems strenous to tackle it will take time. I am still working on crunching the data and possible fixes.
> > 
> > 
> > Yes, but we also now have many Outreachy participants who would be eager to rectify this. They would be good first issues . Rectification could be quicker than you th
> 
> > > At a minimum we should determine a way to make the tests fail when these types of errors occur. The testing regime is therefore flawed.
> > 
> > 
> > Exactly, our test setup is allowing those warning. We should fail the test if those warning appear.
> 
> Actually we need to understand how react is build @anshgoyalevil . Mismatching of variable for same purpose is what it not permits but just throws a warning.

This should have been taken into account while writing tests. This is the sole reason for most of warnings.

**Comment by palisadoes** [comment]
I count:

1. 83 warnings with the `This typically indicates a configuration error in your mocks setup, usually due to a typo or mismatched variable` message
1. They look like they are concentrated on the files 
    1. Roles.tsx
    2. OrganizationEvents.tsx (@aashimawadhwa I think you are working on this now)
    3. OrgList.test.tsx
    4. AdminNavbar.test.tsx
    5. OrgPost.test.tsx
    6. OrgSettings.test.tsx
    7. OrganizationEvents.test.tsx
    8. App.test.tsx
    9. AddOnRegister.test.tsx
    10. OrgContribution.test.tsx
    11. AddOn.test.tsx

This looks like a manageable list for us to create an issue per file and a final one to fail tests on warnings. What do you think?

**Comment by aviraldevv** [comment]
Yes @palisadoes . I and @anshgoyalevil  can volunteer you out with this.

**Comment by palisadoes** [comment]
1. Please start with creating the issues
2. Let the #Talawa-Admin and #general channels slack channels know that they are available.
3. Apply the final fail on warning modification. 

You can try using [GiHub task lists](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists) I've seen @literaleval has used them for tracking issues. You could ask him for advice on this.

**Comment by aviraldevv** [comment]
Sure @palisadoes .

**Comment by aviraldevv** [comment]
Shall we create issues in phases?? What do you think @palisadoes ?? Like 25 issue in first phase then so on.

**Comment by palisadoes** [comment]
Take the initiative. Your choice on the best approach

**Comment by aviraldevv** [comment]
Thanks for giving us the responsibility.

**Comment by aashimawadhwa** [comment]
maybe i can also help you in it @anshgoyalevil @aviraldevv

**Comment by aviraldevv** [comment]
> Take the initiative. Your choice on the best approach.
> 
> I think on about a dozen files are affected.

Yes I was thinking in terms of warning my bad.

**Comment by palisadoes** [comment]
I came to that conclusion by grep-ping and awk-ing the error messages for `tests.txt`. 

`grep test\.tsx ~/Downloads/talawa-admin-errors.txt  | grep Object | awk '{print $3}'`

**Comment by anshgoyalevil** [comment]
@palisadoes Thanks Sir. 

I guess, that is a Unix Command.

**Comment by palisadoes** [comment]
It should work on a Mac. I use kubuntu for all my systems. Windows only occasionally for PowerPoint.

**Comment by palisadoes** [comment]
Happy Holi.

**Comment by palisadoes** [comment]
What are you guys doing? You should outside having fun!

**Comment by aashimawadhwa** [comment]
> 

I am using mac will check in a while and let you know.

**Comment by aviraldevv** [comment]
Happy Holi everyone @palisadoes @anshgoyalevil @aashimawadhwa .

**Comment by anshgoyalevil** [comment]
@palisadoes 

Holi comes and go every year. But, we have 2 years left to level up our career. 🙂

**Comment by palisadoes** [comment]
Dinner time and bed for me. TTYL

**Comment by aviraldevv** [comment]
> Dinner time and bed for me. TTYL

Me who didn't sleep whole night.
